paddle/fluid/framework/tensor_util.cc

```diff
@@ -48,6 +48,10 @@ void TensorCopy(const Tensor& src, const platform::Place& dst_place,
   }
 
   auto size = src.numel() * SizeOfType(src.type());
+#ifdef PADDLE_WITH_MKLDNN
+  if (src.layout() == DataLayout::kMKLDNN)
+    size = std::min(src.memory_size(), dst->memory_size());
```

Comment:
Ideally it should but I introduced `min()` here because `memory_size()` isn't controlled inside TensorCopy so it protects from copying outside of allocation space. You think we should enforce same memory sizes here?

Reply:
I'm not sure if I understand situation here. so what will happen in situation when we Were src 
holding NCHW16C data so numel()*sizeof(data_type) < size and we want dst to to be same size
as src ? dst size seems to be calculated using numel()*sizeof(data_type) , so dst_size < src_size. Is copying data in such a situation introducing a risk of having not everything copied ?

Reply:
Maybe I stated my point not very precisely. Introducing `std::min` here doesn't affect current outcome of this code, it's just safety mechanism (and I'm not sure if it's right approach, maybe it would be better to crash if memory_sizes don't match). What is important here is that for MKL-DNN data layout we can't look at `numel()*sizeof(data_type)` because in blocked format there could be gaps and memory block can be bigger. Both src and dst `.memory_size()` are internal tensor values set earlier by `mutable_data(..., size_t requested_size)` functions. If we omit this `if (src.layout() == DataLayout::kMKLDNN)` we will end up with not copying everything and e.g. models will run 'correctly' but with much less accuracy.

Reply:
dst->memory_size() is returning size of alloction made at dst->mutable_data() which is numel()*sizeof(src.type())  so most of time src.memory_size() and dst->memory_size() should be equal. The only situation when they are not equal is that we src is FP32 and blocked data arrangement is used. Then src.memory_size() is bigger than dst->memory_size().  if src's memory_size() is bigger then copying std::min(src.memory_size(), dst->memory_size()) will get us part of valid data copied? So it dst get corrupted. Should we have std::max(src.memory_size(), dst->memory_size()) ?

Reply:
Allocation of data can be done before calling `TensorCopy()` and this is what is done in `conv_mkldnn_op.cc` in lines `319-320` this PR. In this scenario both src and dst memory_sizes are equal. But you are right, if one of them is greater than the other, we will end up with corrupted data. Either not all data will be copied `std::min()` or we will try write on not allocated memory region `std::max()`. Should I delete this safety mechanism and add for example `PADDLE_ENFORCE_EQ` on memory sizes?

Reply:
I think I understand now. In general I think It would be good to use PADDLE_ENFORCE_EQ , just to catch otherwise difficult to trace problems with copying Tensors when blocked formats are used. Please check if adding this ENFORCE and then setting size to any of src / dst memory_size works fine.